### PR TITLE
Check before handling 404 routes

### LINKED ISSUE #182: 404 handling breaks when no routes for a particular method are defined
If I have the following set of routes:

```php
$router = new \Bramus\Router\Router();

$router->get('/', function() { /* ... */ });

$router->mount('/page-a', function() use (&$router) {
    $router->get('/sub-a', function() { /* ... */ });
    $router->get('/sub-b', function() { /* ... */ });
});

$router->mount('/page-b', function() use (&$router) {
    $router->get('/sub-a', function() { /* ... */ });
    $router->get('/sub-b', function() { /* ... */ });
});

$router->run();
```

then navigating to `/nonexistent-page` in a browser will correctly display the browser's standard 404 page. 

However, if I attempt to send a *POST* request to `/nonexistent-page`, then I get a `200 OK` and see the following error instead:

> **Warning**: Undefined array key "POST" in **C:\Users\Matthew\\...\backend\vendor\bramus\router\src\Bramus\Router\Router.php** on line **292**
> 
> **Warning**: Cannot modify header information - headers already sent by (output started at C:\Users\Matthew\\...\backend\vendor\bramus\router\src\Bramus\Router\Router.php:292) in **C:\Users\Matthew\\...\backend\vendor\bramus\router\src\Bramus\Router\Router.php** on line **374**

However&sup2;, if I set any POST route on the router, i.e.

```php
$router = new \Bramus\Router\Router();

// -- snip

$router->post('/', function() {
    echo 'yeet';
});

$router->run();
```

then all other POST endpoints now correctly respond with 404.

This behavior is also true for `PUT`, `DELETE`, `OPTIONS`, and `PATCH`. Since `HEAD` is handled by a special case, it works just fine and returns 404 -- see [curl-logs.txt](https://github.com/bramus/router/files/8799502/curl-logs.txt).

I doubt that this behaviour is intended. In my eyes, any undefined route/match should get 404'd, i.e.

```php
// This should give me a 404 for every route, but instead it spits out errors.
$router = new \Bramus\Router\Router();
$router->run();
```

If it is intended, and the developer is simply expected to define handlers for all method types, then I guess that's my answer.

---

Per the error message, this issue comes from line 292 (`Router::run()`):

https://github.com/bramus/router/blob/55657b76da8a0a509250fb55b9dd24e1aa237eba/src/Bramus/Router/Router.php#L290-L298

when the router attempts to index into `$this->afterRoutes` using the current method. Since `afterRoutes` starts as `array()`, there are no keys for **any** method to start, which makes this an index out of range call.

Funny to note that, further down at line 371 (`Router::trigger404()`), there is actually an `isset` check before attempting to invoke the `notFoundCallback` for `'/'`:

https://github.com/bramus/router/blob/55657b76da8a0a509250fb55b9dd24e1aa237eba/src/Bramus/Router/Router.php#L371-L375

With that in mind, there are two ways that I've found to fix this: 

1.  Adding a check to line 292 before invoking:

    ```php
    $match = $this->afterRoutes[$this->requestedMethod] ?? '/';
    $this->trigger404($match);
    // Probably `if (isset())` instead of `??` for backwards compat.
    ```

2.  Giving the routes' method keys default values:

    ```php
    /**
     * @var array The route patterns and their handling functions
     */
    private $afterRoutes = array(
      'GET' => array(),
      'POST' => array(),
      'PUT' => array(),
      'DELETE' => array(),
      'OPTIONS' => array(),
      'PATCH' => array(),
      'HEAD' => array()
    );

    /**
     * @var array The before middleware route patterns and their handling functions
     */
    private $beforeRoutes = array(
      'GET' => array(),
      'POST' => array(),
      'PUT' => array(),
      'DELETE' => array(),
      'OPTIONS' => array(),
      'PATCH' => array(),
      'HEAD' => array()
    );
    ```

I would love to submit a pull request with either of these changes for further review, since I've already got it started. It should probably receive a bit more thorough testing.

Cheers,
-- Matt

### PR DESCRIPTION
**Pull request by matthew-e-brown**
This PR fixes #182.

I have verified that PHPUnit tests still pass. On a side-note, the version of PHP unit in `composer.json` doesn't work with PHP 8 due to the deprecation of `each` that it uses. Might be worth updating it.

### CODE REVIEW

File: src/Bramus/Router/Router.php [hunk 1]
```diff
@@ -288,13 +288,11 @@ public function run($callback = null)
...
         // If no route was handled, trigger the 404 (if any)
-        if ($numHandled === 0) {
+        if ($numHandled === 0 && isset($this->afterRoutes[$this->requestedMethod])) {
             $this->trigger404($this->afterRoutes[$this->requestedMethod]);
         } // If a route was handled, perform the finish callback (if any)
-        else {
-            if ($callback && is_callable($callback)) {
-                $callback();
-            }
+        elseif ($callback && is_callable($callback)) {
```
Comment (bramus) [inline src/Bramus/Router/Router.php:294]: By introducing the `elseif`, it also introduces the unwanted side-effect of running the callback when there is no handler and no afterRoute … To not change the existing behavior, the callback should only be run when there was a handler.

Comment (bramus) [reply src/Bramus/Router/Router.php:294]: The solution lies at line 291: don’t `&&` the extra condition, but nest it. Then this `elseif` could remain in place.

**Review by bramus — CHANGES_REQUESTED** [review]

**Review by bramus — APPROVED** [review]

**Comment by matthew-e-brown** [comment]
If my investigation is correct, it looks like Git's [recent-ish update to fix CVE-2022-24765](https://github.blog/2022-04-12-git-security-vulnerability-announced/) is making this action fail. Perhaps updating to `v2.4.1` as per actions/checkout#766 would fix it? 🤔

**Comment by bramus** [comment]
Good find! It took me some attempts, but I have fixed the GitHub Actions build on `master`. If you could merge, then this PR can be tested as well.

**Comment by matthew-e-brown** [comment]
Just pulled your `master` changes into my fork. Is that what you meant by merge? 😅

**Pull request merged**